p2p/src/main/kotlin/maru/p2p/P2PNetworkFactory.kt

```diff
@@ -0,0 +1,165 @@
+import pubsub.pb.Rpc
+import tech.pegasys.teku.infrastructure.async.AsyncRunnerFactory
+import tech.pegasys.teku.infrastructure.async.MetricTrackingExecutorFactory
+import tech.pegasys.teku.infrastructure.unsigned.UInt64
+import tech.pegasys.teku.networking.p2p.libp2p.LibP2PNetwork
+import tech.pegasys.teku.networking.p2p.libp2p.LibP2PNodeId
+import tech.pegasys.teku.networking.p2p.libp2p.PeerManager
+import tech.pegasys.teku.networking.p2p.libp2p.gossip.GossipTopicHandlers
+import tech.pegasys.teku.networking.p2p.libp2p.gossip.LibP2PGossipNetwork
+import tech.pegasys.teku.networking.p2p.libp2p.gossip.PreparedPubsubMessage
+import tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler
+import tech.pegasys.teku.networking.p2p.network.P2PNetwork
+import tech.pegasys.teku.networking.p2p.network.PeerHandler
+import tech.pegasys.teku.networking.p2p.peer.Peer
+import tech.pegasys.teku.networking.p2p.reputation.ReputationManager
+
+object P2PNetworkFactory {
+  fun build(
+    privateKey: PrivKey,
+    port: String,
+    ipAddress: String,
+  ): P2PNetwork<Peer> {
+    val addr = Multiaddr("/ip4/$ipAddress/tcp/$port")
+
+    return setupP2PNetwork(privateKey, addr)
+  }
+
+  private fun setupP2PNetwork(
+    privateKey: PrivKey,
+    ipv4Address: Multiaddr,
+  ): P2PNetwork<Peer> {
+    val rpcMethod = MaruRpcMethod()
+    val gossipTopicHandlers = GossipTopicHandlers()
+//    gossipTopicHandlers.add("topic", TestTopicHandler()) // TODO: I don't think this is needed
+
+    val gossipParams = GossipParamsBuilder().heartbeatInterval(100.millis).build()
+    val gossipRouterBuilder =
+      GossipRouterBuilder().apply {
+        params = gossipParams
+        scoreParams = GossipScoreParams(GossipPeerScoreParams(), GossipTopicsScoreParams())
+        messageFactory = { getMessageFactory(it, gossipTopicHandlers) }
+      }
+    val gossipRouter = gossipRouterBuilder.build()
+    val pubsubApiImpl = PubsubApiImpl(gossipRouter)
+    val gossip = Gossip(gossipRouter, pubsubApiImpl)
+
+    val metricsSystem = NoOpMetricsSystem()
+    val publisherApi = gossip.createPublisher(privateKey, Random.nextLong())
+    val gossipNetwork = LibP2PGossipNetwork(metricsSystem, gossip, publisherApi, gossipTopicHandlers)
+
+    val peerId = PeerId.fromPubKey(privateKey.publicKey())
+    val libP2PNodeId = LibP2PNodeId(peerId)
+
+    val metricTrackingExecutorFactory = MetricTrackingExecutorFactory(metricsSystem)
+    val asyncRunner = AsyncRunnerFactory.createDefault(metricTrackingExecutorFactory).create("maru", 2)
+    val rpcHandler = RpcHandler(asyncRunner, rpcMethod)
+
+    val peerManager =
+      PeerManager(
+        metricsSystem,
+        ReputationManager.NOOP,
+        mutableListOf<PeerHandler>(MaruPeerHandler()),
+        mutableListOf(rpcHandler),
+        { _ -> 50.0 }, // TODO: I guess we need a scoring function here
+      )
+
+    val host = createHost(privateKey, listOf(gossip, peerManager), gossip, rpcHandler, ipv4Address)
```

Comment:
Consider using named arguments for such long argument lists

Reply:
done